### PR TITLE
[#3780] Use JLayout to render debug backtrace and add backtrace in isis template error.php

### PR DESCRIPTION
**Pull request by asika32764**
Pull Request re-create for #3780 .

## Updated 2017

- Added `JHtmlDebug` helper to render backtrace.
- Also add xdebug link support similar to https://github.com/joomla/joomla-cms/pull/10306 did.
- JError::renderBacktrace() use JHtmlDebug too

-------

## Introduction

In current time, we use `setError()`, `getError()` and `JError::raiseError()` in component to send error message.

However, if we throw `Exception`, the debug plugin will not render backtrace for developer. It only work when we use `JError::raiseError()`.

It's because we have this code in: https://github.com/joomla/joomla-cms/blob/master/plugins/system/debug/debug.php#L213

``` php
if (JDEBUG)
        {
            if (JError::getErrors())
            {
                $html[] = $this->display('errors');
            }
```

To detect whether display errors or not. When we throw an exception, the `JError::getErrors()` will return null that backtrace will not rendered.
## The Modification

I added backtrace layout in `layouts/error/backtrace.php` and now, debug plugin & isis template both include this layout. It will help developers know how errors occurred.
## How to Test
1. Go to `Global Configuration`, enable the debug mode.
2. Then type `index.php?option=foo` in browser, you will see the error page.
## Screenshot

p2013-12-01-3

p-2014-06-14-3

### CODE REVIEW

File: libraries/cms/html/debug.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	/**
+	 * Render backtrace table by JLayout.
+	 *
+	 * @param   array  $backtrace  The backtrace array from exception or debug_backtrace() function.
+	 *
+	 * @return  string  The backtrace table HTML.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function backtrace($backtrace)
```
Comment (mbabker) [inline libraries/cms/html/debug.php]: Personally, I'm not a fan of JHtml functions that are only proxies to render a JLayout.  As JHtml is inherently overridable (and yes I'm aware that override mechanism is more for coders than end users), either the markup should be inlined into the JHtml function or users of this should just call JLayout directly and this function removed.

Comment (asika32764) [inline libraries/cms/html/debug.php]: ok I can remove this method. Is it good idea to just create one method for new class `JHtmlDebug` or we should put `xdebuglink()` in other class?

Comment (mbabker) [inline libraries/cms/html/debug.php]: That method's fine here for now.  I'd love to [expand the debug functionality](https://github.com/joomla/joomla-cms/issues/11840) so it isn't so tightly coupled to the debug plugin, eventually that's either going to mean a library level JHtml helper or a debug library with some sort of helper classes (or possibly both).

**Comment by asika32764** [comment]
Done

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on c7ee7b6b912a43f7649c4d9eeb347585a597bf40<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14957">issues.joomla.org/tracker/joomla-cms/14957</a>.</sub>

**Review by mbabker — APPROVED** [review]

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on c7ee7b6b912a43f7649c4d9eeb347585a597bf40<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14957">issues.joomla.org/tracker/joomla-cms/14957</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful testes.

**Pull request merged**